### PR TITLE
Typescriptify accounts

### PR DESCRIPTION
**Pull request by punmechanic**
WIP

Originally I wanted to convert the React components within `features/accounts` to TypeScript as per the desired guidelines outlined by @renchap in #24843, but I realized that that was folly without first converting the associated redux code to also use TypeScript.

The goal of this PR is to convert that entire slice + the associated redux code to TypeScript

### CODE REVIEW

**Comment by trinitroglycerin** [comment]
Ah, I meant to submit this as a draft, apologies

**Comment by nschonni** [comment]
You could reopen this and then convert it to a draft https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft instead of opening another PR

**Comment by trinitroglycerin** [comment]
Thanks @nschonni; on Github enterprise, the "close pull request button" is where I remember the conversion to draft being, so I didn't even think to look in the reviewer list

**Review by renchap — COMMENTED** [review]
Thanks for your contribution, and I am happy to see more people caring about improving the frontend and moving to Typescript!

You can have a look at the [`typescript` label](https://github.com/mastodon/mastodon/pulls?q=is%3Apr+-label%3Adependencies++label%3Atypescript) to see what has been done so far and where we are moving.

As you saw when working on this code, a big problem at the moment is that there are no `Record` used anywhere, but plain `Map` and `List`, which do not map well at all with typescript (see https://github.com/mastodon/mastodon/pull/25137#discussion_r1242660945 for related discussion).

I started laying down a plan to tackle this, but did not had time in the last weeks to finish and publish it. Maybe this will motivate me :)

My vision for the final state of Redux in Mastodon is:
- it should not contain state that should not be local (example: at the moment, `blocks:` is only used in one specific modal, it can be a local state)
- use RTK patterns everywhere
- drop Immutable completely
- plain Javascript state, using Immer (this comes with RTK)
- fully typed state
- collocate actions & reducers, as they are very tied, they should not be in 2 different sub-directories
- (maybe) introduce a `models/` directory, which will contain both the type definitions for our models (= objects that appear in the state) and any utility function related to those models

The path to get there is complex, and I started reviewing reducers one-by-one to check what should be done with each:
- drop it (yes, there are unused reducers…)
- convert it to local state (or a global JS variable in 1 instance)
- transform it to a plain JS object, use RTK functions, when it is not used in many places and the conversion can be easy. See #25585 for an example
- keep it with Immutable because it is called by a lot of different things that rely on Immutable methods, but transform it into a `Record`. Second step is moving the actions & reducer to RTK, and the last step would be to convert it to plain JS object, but ensuring it is correctly typed first would really help avoid breakages

You focused on `accounts`, and it fell in my last category. It was maybe not the easiest reducer to start with, especially because an `Account` also has `Emoji`, which may need to be properly typed first.

Does this all make sense to you? I would be very happy to have some feedback, and I will try to find a way to better coordinate efforts around Typescript in the coming days (or weeks? 🤐).

And thanks again for your work, it is really appreciated!

File: app/javascript/mastodon/reducers/accounts.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import { Map as ImmutableMap, fromJS } from 'immutable';
+
+import { ACCOUNT_REVEAL } from 'mastodon/actions/accounts';
+import { ACCOUNT_IMPORT, ACCOUNTS_IMPORT } from 'mastodon/actions/importer';
+import type { Account as StateAccount } from 'mastodon/initial_state';
+
+// TODO(trinitroglycerin): the value of this map is actually a record type,
```
[the remark this answers is not in the document]
Comment (trinitroglycerin) [reply app/javascript/mastodon/reducers/accounts.ts:7]: Sorry, when I said "record" here, I don't mean a literal record as in `Immutable.Record`. I mean a record in the computer science definition of the term.

File: app/javascript/mastodon/reducers/accounts.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+const normalizeAccounts = (state: State, accounts: Account[]): State => {
+  accounts.forEach((account) => {
+    state = normalizeAccount(state, account);
+  });
+
+  return state;
+};
+
+type Action =
```
[the remark this answers is not in the document]
Comment (trinitroglycerin) [reply app/javascript/mastodon/reducers/accounts.ts:56]: Potentially. This is a draft PR, I'm taking things one step at a time. Just changing things to be typed correctly without impacting logic is a brain bender :)

File: app/javascript/mastodon/selectors/accounts.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import type { Map as ImmutableMap } from 'immutable';
```
[the remark this answers is not in the document]
Comment (trinitroglycerin) [reply app/javascript/mastodon/selectors/accounts.ts:1]: The motivation behind this was not because it's nicer (it is), but because it appears importing anything from a js file causes my vscode linter to assume the type is `any`, even when it's not :(

File: app/javascript/mastodon/reducers/accounts.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  statuses_count: number;
+}
+
+// This complicated type is used to allow for the deletion of the properties in Counters from the Account type.
+//
+// The properties in the Counters type are derived values and are tracked in the account_counters slice.
+//
+// TODO(trinitroglycerin): This additionally adds the derived hidden and limited properties, as these are not present
+// in the jsdoc types. However, there are a few instances where React components rely on these properties existing,
+// so it's possible that the hidden and limited properties should be graduated to the jsdoc types.
```
[the remark this answers is not in the document]
Comment (trinitroglycerin) [reply app/javascript/mastodon/reducers/accounts.ts:29]: Yeah, what I have atm (I'll push it up in a second) is a hack type:

```ts
interface TypeSafeImmutableMapImpl<T extends object> {
  get<K extends keyof T, V extends T[K] | null | undefined = undefined>(
    key: K,
    notSetValue?: V
  ): V;
}

export type TypeSafeImmutableMap<T extends object> =
  TypeSafeImmutableMapImpl<T> & Map<keyof T, unknown>;


// TODO(trinitroglycerin): This is NOT typed correctly for nested collections.
//
// fromJS(T) will take a T and turn any nested collections into List() or Map() as appropriate, however
// the typing of this function implies that arrays and objects are untouched.
//
// This could be fixed using Pick and generics, maybe.
export function intoTypeSafeImmutableMap<T extends object>(
  t: T
): TypeSafeImmutableMap<T> {
  return fromJS(t);
}
```
This provides type safety at the first level - the following compiles (if `state.accounts` were `Immutable.Map<string, unknown`, it would not):

```ts
const getAccountBase = (
  state: RootState,
  id: string
): TypeSafeImmutableMap<Account> | null => {
  return state.accounts.get(id, null);
};
```

However, there are some patterns that are not possible to make type safe. Notably, the use of `updateIn()` in immutable isn't type safe - it's not possible to type this in typescript without assertions:

```ts
const incrementFollowers = (state: State, accountId: string): State => {
  state
    .updateIn([accountId, 'followers_count'], (num) => (num as number) + 1)
    .updateIn([me, 'following_count'], (num) => (num as number) + 1);
};
```

File: app/javascript/mastodon/reducers/accounts.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+  return state;
+};
+
+type Action =
+  | { type: typeof ACCOUNT_IMPORT; account: Account }
+  | { type: typeof ACCOUNTS_IMPORT; accounts: Account[] }
+  | { type: typeof ACCOUNT_REVEAL; id: unknown };
+
+export function accounts(state = initialState, action: Action) {
```
[the remark this answers is not in the document]
Comment (trinitroglycerin) [reply app/javascript/mastodon/reducers/accounts.ts]: Done

File: app/javascript/mastodon/reducers/accounts.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (trinitroglycerin) [reply app/javascript/mastodon/reducers/accounts.ts]: with regards to the usage of `updateIn`, I think this is a case where RTK/using POJOs with immer helps a lot.

**Comment by trinitroglycerin** [comment]
I didn't see that related PR. There's a significant overlap between what I am doing and what the other author is doing. Interestingly, I only started typing the accounts redux module so that I could type the components :)

I may drop this effort and focus on a different segment of the codebase, in that case.

File: app/javascript/mastodon/reducers/accounts.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (renchap) [reply app/javascript/mastodon/reducers/accounts.ts]: Yes, unfortunately the `*In` methods in Immutable can not be typed at all :(

This is why moving to plain JS objects + Immer is inevitable if we want to be fully typed.

But `Record` allow you to use dot accessors, so `value.getIn('p1', 'p2', 'p3')` can be written at `value.p1.p2.p3`. This does not solve `updateIn`, but helps in many places.

**Comment by renchap** [comment]
> I only started typing the accounts redux module so that I could type the components :)

Yes, many of the components not relying on the state have already been converted, but now the state is the next logical step.

I worked around this in a few PRs but this is not ideal:
- [Using "fake" types with only the fields needed by this specific code ](https://github.com/mastodon/mastodon/pull/26499/files#diff-a5c561a3faa113252bf282ebf33eb113b3329d30ba3b5fc4ca4952f79bdd047eR18)
- [Working with `any` and ignoring `eslint` error for the file](https://github.com/mastodon/mastodon/pull/25583/files#diff-f482938ef7dae4bfc7279bb73f4bbea6879da2a85ad51ebbf1e9a4da8bda836eR1)

I would not encourage going the same way for more components, those will probably bite me in the future.

> I may drop this effort and focus on a different segment of the codebase, in that case.

Sorry for this, I do not want to prevent you for contributing!

I am trying to outline what the roadblocks can be on this matter, as I have been thinking on the best way to move forward on this topic for some time.

You motivated me to write down everything I had in my mind in #26556 and I hope we will be able to work out a way to move forward on typing the `accounts` state :)

**Comment by trinitroglycerin** [comment]
You haven't at all prevented me from contributing :) I'll keep hacking away and maybe we can find a space where we merge the two pull requests together. After all, the other pull request doesn't touch the components, which I'm currently doing. Since the type changes in theory shouldn't impact the public API of Redux, the contribution there is still useful

**Comment by renchap** [comment]
I could not stop thinking about it, so I went ahead and had a first try at using proper `Record` for accounts: https://github.com/mastodon/mastodon/pull/26559

Is it not ready yet (I saw that you noticed some fields that are not documented), but it should give a good idea of what is involved.

What do you think, compared to using your custom type? It is a bit more risky as `Record`s are stricter than `Map`s (unknown properties are dropped on `set` for example), but it looks cleaner to me.

**Comment by trinitroglycerin** [comment]
@renchap For as long as Immutable is used I think the Record is a better approach than my type hacks. The nice thing about the type is that it has no runtime cost, but I'm also really not confident in it being 100% accurate, and having to write out all of the functions you want access to is rough. We should go with your approach.

If you get that merged in, I'll rebase my code off of it so it's just the UI changes.

**Comment by trinitroglycerin** [comment]
@renchap Merged your changes in, they definitely look easier to maintain. I'll keep cherry-picking them in as they come

**Comment by renchap** [comment]
You really did not pick the easiest part of the codebase. I spent a few hours today on this, and this is really complex because it touches multiple other reducers, like `relationships` and `accounts_counters` (which should probably be merged in the `accounts` reducer, I dont see a reason it is separate).

Those are all mixed together in `makeGetAccount` which creates the final object that is used by all the components, so if you want to type the components, you need to correctly type all of those, and each one is a new can of worms I discover.

I will continue working on this, but I am making no promises on a timeframe. I am also a bit afraid that I will break things while I go, as it touches more and more areas.

You can continue working on your components conversion, but it wont be merged until I am done with all the Redux side.
Also I think at the end it will be better if you split the work into multiple PRs with a few components in each, so we can test and review them independently.

And note that we only have a few tests (ie, two in total) [covering the frontend](https://github.com/mastodon/mastodon/blob/main/spec/system/new_statuses_spec.rb), so this is refactoring untested code with a lot of edge-cases (the `Relationship` model for example).

**Comment by trinitroglycerin** [comment]
Haha, if you take a look at the files I've changed I actually encountered all of those things :)

* `relationships` I have yet to touch
* `account_counters` is typed and using records as of 8476040, which you can probably cherry-pick to save yourself some trouble
* `makeGetAccount` is correctly typed in this PR
* As for tests, the components I've touched I've not modified the logic of except for `AccountNote`, which I added tests for, then refactored and ensured those tests still passed :) I think I added like 20% additional codebase tests, lol

**Comment by renchap** [comment]
> * `relationships` I have yet to touch
I am having a try at it. FYI, you can find all the fields that are in the API JSON in Rails's serializers, for example https://github.com/mastodon/mastodon/blob/main/app/serializers/rest/relationship_serializer.rb

I am using this to define the `api_types`, and they trying to find out how and where the other fields come from, if any.



> * `account_counters` is typed and using records as of [8476040](https://github.com/mastodon/mastodon/commit/8476040b307b092feb987d9c2b785887e257d62d), which you can probably cherry-pick to save yourself some trouble

Maybe you could extract this in another PR focusing on converting `account_counters`? I really prefer small and incrementals PRs, and those changes should be easy to extract



> * `makeGetAccount` is correctly typed in this PR

I just saw this, nice work!

What do you think of also writing a full custom PropType definition (see [`recordOf` example here](https://github.com/HurricaneJames/react-immutable-proptypes))  for your `ExposedAccount`, and replacing it in every components that uses `makeAccountSelector`? It might make it easier to catch any field that is missing in the `Record` conversion, and might make us more confident when moving to TS?



> * As for tests, the components I've touched I've not modified the logic of except for `AccountNote`, which I added tests for, then refactored and ensured those tests still passed :) I think I added like 20% additional codebase tests, lol

Nice work here as well, I missed this when I glanced at your recent changes. There are not a lot of Jest React tests in the code base, and I personally do not have much experience with them. If you want real end-to-end tests, you can have a look at the system specs I introduced recently (#25461). It has been on my todolist to add more and to cover both the main flows, and the very obscure places (like the moderation interface or the share URL, which use React in maybe not expected ways).

Also, maybe you could extract those tests in a separate PR if they do not rely on your TS migration?

**Comment by trinitroglycerin** [comment]
The tests I wrote unfortunately do depend on some changes to `AccountNote`, but I've raised a separate PR for those tests + those changes: #26570
